src/main/java/org/opensearch/knn/indices/ModelIndex.java

```diff
@@ -0,0 +1,231 @@
+import org.opensearch.action.get.GetResponse;
+import org.opensearch.action.index.IndexRequestBuilder;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.action.support.WriteRequest;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.util.KNNEngine;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_INDEX_MAPPING_PATH;
+import static org.opensearch.knn.common.KNNConstants.MODEL_INDEX_NAME;
+import static org.opensearch.knn.index.KNNSettings.MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING;
+import static org.opensearch.knn.index.KNNSettings.MODEL_INDEX_NUMBER_OF_SHARDS_SETTING;
+
+/**
+ * ModelIndex is a singleton class that controls operations on the model system index
+ */
+public final class ModelIndex {
+    public static Logger logger = LogManager.getLogger(ModelIndex.class);
+
+    private int numberOfShards;
+    private int numberOfReplicas;
+
+    private static ModelIndex INSTANCE;
+    private static Client client;
+    private static ClusterService clusterService;
+    private static Settings settings;
+
+    /**
+     * Make sure we just have one instance of model index
+     *
+     * @return ModelIndex instance
+     */
+    public static synchronized ModelIndex getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new ModelIndex();
+        }
+        return INSTANCE;
+    }
+
+    public static void initialize(Client client, ClusterService clusterService, Settings settings) {
+        ModelIndex.client = client;
+        ModelIndex.clusterService = clusterService;
+        ModelIndex.settings = settings;
+    }
+
+    private ModelIndex() {
+        numberOfShards = MODEL_INDEX_NUMBER_OF_SHARDS_SETTING.get(settings);
+        numberOfReplicas = MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_SHARDS_SETTING,
+                it -> numberOfShards = it);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING,
+                it -> numberOfReplicas = it);
+    }
+
+    /**
+     * Creates model index. It is possible that the 2 threads call this function simulateously. In this case, one
+     * thread will throw a ResourceAlreadyExistsException. This should be caught and handled.
+     *
+     * @param actionListener CreateIndexResponse listener
+     * @throws IOException thrown when get mapping fails
+     */
+    public void create(ActionListener<CreateIndexResponse> actionListener) throws IOException {
+        if (isCreated()) {
```

Comment:
Should we log message here and return?

Reply:
I am a little worried this may blow up the logs if isCreated is not called by caller. Id prefer not to log here.